### DISCUSSION TITLE
New API for Direct Messaging between network participants

### DISCUSSION THREAD

**Discussion by ravi-prakash-v** [beckn/protocol-specifications#270] [Proposals]
There have been a couple of requests made by members of the DHP community to introduce a `message` API that allows a BAP to send a DM to the BPP and vice-versa.

It did not make sense to create an `on_message` callback API as messages are independent. Hence just a `message` API.

Below is an inital draft of the `message` API. 

```
{
    "context" : {
        "domain" : "<domain_code>",
        "country" : "IND",
        "city" : "std:011",
        "action" : "message",
        "message_id" : "<uuid>",
        "transaction_id" : "<uuid>",
        "timestamp" : "<date time>",
        "ttl" : "100"
    },
    "messsage" : {
        "text" : "Quick brown fox",
        "data" : {
            "mimetype" : "image/png",
            "value" : "<data>"
        },
        "vcard" : "<vcard formatted string>",
        "location" : {
            "gps" : "11.87454,77.987398"
        } 
    }
}
```

This not only has text messaging but also allows sharing of files, contacts and location. 

Thoughts?

**Comment by ravi-prakash-v** [comment]
The `message_id` will be constant during a chat session and `transaction_id` can link the chat session for a specific order.

**Reply by manojcraftsvilla** [reply]
Thanks Ravi. With this architecture both DM and chats can be built. Hopefully this will also come in ONDC since chat is very much needed in an ecommerce transaction.

**Comment by manojcraftsvilla** [comment]
My one observation is regarding "reply to" feature which is a reply to a specific message (like in Whatsapp) in a chat session. Usually a message_id and thread_id would have helped in making that happen. One more requirement in chat which might come later is a CTA button with URL to enable click actions in chat

**Reply by ravi-prakash-v** [reply]
@manojcraftsvilla Adding a CTA is terrific idea! Can you modify the above json and paste it below?

**Comment by manojcraftsvilla** [comment]
Below is attempt to include CTA (Call to Action) options in message. Three CTA types can be used just like in Whatsapp: Button, List of choices and Quick reply (eg. Yes/No or Pickup/Home Delivery). ctatext therefore is an array below. For button type, only first element of array will be used as Button text

{
    "context" : {
        "domain" : "<domain_code>",
        "country" : "IND",
        "city" : "std:011",
        "action" : "message",
        "message_id" : "<uuid>",
        "transaction_id" : "<uuid>",
        "timestamp" : "<date time>",
        "ttl" : "100"
    },
    "messsage" : {
        "text" : "Quick brown fox",
        "cta" : {
            "ctatype": "button, list or quickreply",
            "ctatext" : ["Pay Now"],
            "ctalink" : "https://upi.pay/?phone=9999999999"
         },   
        "data" : {
            "mimetype" : "image/png",
            "value" : "<data>"
        },
        "vcard" : "<vcard formatted string>",
        "location" : {
            "gps" : "11.87454,77.987398"
        } 
    }
}

**Comment by manojcraftsvilla** [comment]
Done

Manoj

**Comment by ravi-prakash-v** [comment]
Re-pasting the JSON in prettified format

```
{
    "context" : {
        "domain" : "<domain_code>",
        "country" : "IND",
        "city" : "std:011",
        "action" : "message",
        "message_id" : "<uuid>",
        "transaction_id" : "<uuid>",
        "timestamp" : "<date time>",
        "ttl" : "100"
    },
    "messsage": {
        "text": "Quick brown fox",
        "cta": {
            "ctatype": "button, list or quickreply",
            "ctatext": [
                "Pay Now"
            ],
            "ctalink": "https://upi.pay/?phone=9999999999"
        },
        "data": {
            "mimetype": "image/png",
            "value": ""
        },
        "vcard": "<vcard formatted string>",
        "location": {
            "gps": "11.87454,77.987398"
        }
    }
}
```

**Reply by pramodkvarma** [reply]
@core-wg-admin I suggest we also look at message format specifications (not routing/auth/connection aspects, just message format specifications) of existing chat protocols such as [XMPP](xmpp.org).

**Reply by pramodkvarma** [reply]
@core-wg-admin basically messaging in beckn is between BAP and BPP for a transaction context. Gor example anyone from seller side should be able to respond instead of addressing directly to a user handle. BAP/BPPs have to take the incoming message and if there is a user ID (specific to their platform) they need to show the message to that user. 

So, basic elements in message exchange on beckn:
1. Message itself: `text` (should it be a markdown or pure text? Should be able to contain links and emojis, using HTML is risky due to script injection, so a markdown structure allows basic formatting, links, etc), 
2. Attachment: `link` and `mime-type`
3. Reference: `parent message ID` (when replying to earlier message)
4. User info (from/to): `user display name` and `internal user id` (of BAP/BPP)

I don't think we need vcard (no need to handle contact management like chat apps), also attachment can handle standard attachments and mime types. I also don't think we need action button separately given URLs can be passed in message text (using markdown format).

**Comment by manojcraftsvilla** [comment]


**Reply by ravi-prakash-v** [reply]
Removing this message as it contains promotional content.

**Comment by ravi-prakash-v** [comment]
Re-pasting the above message without promotional content. 

Pasting an XMPP message format with image sharing and html for reference:
```
<message ***@***.***' ***@***.***'>
    <body>Look at the nice view from the summit.</body>
    <html xmlns='http://jabber.org/protocol/xhtml-im'>
      <body xmlns='http://www.w3.org/1999/xhtml'>
        <p>Look at the nice <p style='font-weight:bold;
display:inline">view</p> from the summit.</p>
        <img
src="ni:///sha3-256;wqfDv8OGw7jCvx7Dl2ZRw4FHVsKgYcOWYsO14oKsw79Nw6Q7ScO64oCaw5gKS-KCrMO4Q0o"
/>
      </body>
    </html>
    <reference xmlns='urn:xmpp:reference:0' begin='17' end='20' type='data'>
      <media-sharing xmlns='urn:xmpp:sims:1'>
        <file xmlns='urn:xmpp:jingle:apps:file-transfer:5'>
          **<media-type>image/jpeg</media-type>
          <name>summit.jpg</name>**
          <size>3032449</size>
          <hash xmlns='urn:xmpp:hashes:2'
algo='sha3-256'>2XarmwTlNxDAMkvymloX3S5+VbylNrJt/l5QyPa+YoU=</hash>
          <hash xmlns='urn:xmpp:hashes:2'
algo='id-blake2b256'>2AfMGH8O7UNPTvUVAM9aK13mpCY=</hash>
          <desc>Photo from the summit.</desc>
          <thumbnail xmlns='urn:xmpp:thumbs:1'uri='
***@***.***'
media-type='image/png' width='128' height='96'/>
        </file>
        <sources>
          <reference xmlns='urn:xmpp:reference:0' type='data' uri='
https://download.montague.lit/4a771ac1-f0b2-4a4a-9700-f2a26fa2bb67/summit.jpg'
/>
          <reference xmlns='urn:xmpp:reference:0' type='data'
***@***.***/resource?jingle-ft' />
        </sources>
      </media-sharing>
    </reference>
  </message>
```
